test/e2e_node/image_list.go

```diff
@@ -151,27 +152,39 @@ func PrePullAllImages() error {
 	}
 	images := framework.ImageWhiteList.List()
 	klog.V(4).Infof("Pre-pulling images with %s %+v", puller.Name(), images)
+
+	var wg sync.WaitGroup
+	var once sync.Once
+	wg.Add(len(images))
 	for _, image := range images {
-		var (
-			err    error
-			output []byte
-		)
-		for i := 0; i < maxImagePullRetries; i++ {
-			if i > 0 {
-				time.Sleep(imagePullRetryDelay)
+		go func(image string) {
```

Comment:
One thing I was thinking about when I originally filled the issue - do we want to limit the number of images that we try and pull at once?

In other words, if `images` grows to be long in length, could we envision it becoming too expensive to try and download all of them in parallel? How could we protect ourselves against that happening?

Reply:
At this moment，the images length is 20+，so this may not be expensive？other words，we can fix this when it happenes。

If we need fix it now，how about use ** semaphore** to limit the number of goroutines ？

Reply:
Hmmm, I feel like while we are working on this code, we might as well make it robust for the future? What do you think?

Semaphores is definitely one path to consider. I'd also wonder about [Worker pools](https://gobyexample.com/worker-pools). This allows us to leverage channels which I think is a _slightly_ more idiomatic way of doing concurrency in Golang.

Reply:
Agree with you that we can make it robust as well as we can. I will work on it with **Worker pools** support.

Reply:
Awesome, thank you :)

Reply:
done, PTAL